### PR TITLE
add ed25519 sign

### PR DESCRIPTION
**Pull request by liuzemei**

### CODE REVIEW

File: developers/src/assets/js/tools.js [hunk 1]
```diff
@@ -114,20 +130,57 @@ function signEncryptedPin(pin, pinToken, sessionId, privateKey, iterator) {
...
+  return digest.slice(0, 32)
+}
+
 function hexToBytes(hex) {
-  const bytes = []
+  const bytes = new Uint8Array(32)
+  let i = 0
   for (let c = 0; c < hex.length; c += 2) {
-    bytes.push(parseInt(hex.substr(c, 2), 16))
+    bytes[i++] = parseInt(hex.substr(c, 2), 16)
```
Comment (jadeydi) [inline developers/src/assets/js/tools.js:183]: Why not use c / 2 for i?

File: developers/src/components/UpdateToken.vue [hunk 2]
```diff
@@ -65,14 +65,15 @@
...
+    const { session_id, pin_token,private_key } = this.submit_form
+    if (!validator.isUUID(session_id, 4)) {
       this.$message.error({
         message: this.$t("message.errors.session_id_format"),
         showClose: true
       })
       return false
     }
-    if (!validator.isBase64(this.submit_form.pin_token)) {
+    if (!validator.isBase64(pin_token) && Buffer.from(pin_token, 'base64').length !== 32) {
```
Comment (jadeydi) [inline developers/src/components/UpdateToken.vue:76]: We should prefer to use base64 raw URL  pin_token_base64.

File: developers/src/assets/js/tools.js [hunk 3]
```diff
@@ -114,20 +130,57 @@ function signEncryptedPin(pin, pinToken, sessionId, privateKey, iterator) {
...
+function signEncryptEd25519PIN(pinToken, privateKey) {
+  privateKey = Buffer.from(privateKey, 'base64')
+  pinToken = Buffer.from(pinToken, 'base64')
+  return scalarMult(privateKeyToCurve25519(privateKey), pinToken.slice(0, 32))
+}
+
+function scalarMult(curvePriv, publicKey) {
+  curvePriv[0] &= 248
+  curvePriv[31] &= 127
+  curvePriv[31] |= 64
```
Comment (jadeydi) [inline developers/src/assets/js/tools.js:162]: Do these operations still need for curve private?

File: developers/src/assets/js/tools.js (hunk 1 above)
Comment (liuzemei) [reply developers/src/assets/js/tools.js]: fixed

File: developers/src/assets/js/tools.js [hunk 4]
```diff
@@ -1,9 +1,9 @@
-import moment from 'moment'
 import forge from 'node-forge'
 import jwt from 'jsonwebtoken'
 import { v4 as uuid } from 'uuid'
-import LittleEndian from "int64-buffer"
+import { Int64BE } from "int64-buffer"
```
Comment (jadeydi) [inline developers/src/assets/js/tools.js:4]: This should be Uint64LE

File: developers/src/assets/js/tools.js [hunk 5]
```diff
@@ -70,42 +70,58 @@ function environment() {
...
-  const md = forge.md.sha256.create()
-  md.update(method + uri + body, 'utf8')
-  const payload = {
+  if (uri.startsWith('https://api.mixin.one')) uri = uri.replace('https://api.mixin.one', '')
+  if (uri.startsWith('https://mixin-api.zeromesh.net')) uri = uri.replace('https://mixin-api.zeromesh.net', '')
+  method = method.toLocaleUpperCase();
+  if (typeof (body) === "object") body = JSON.stringify(body);
+  let issuedAt = Math.floor(Date.now() / 1000)
+  let md = forge.md.sha256.create();
+  let _privateKey = toBuffer(privateKey, 'base64');
```
Comment (jadeydi) [inline developers/src/assets/js/tools.js:79]: Private key should use RAWURLEncoding like pin_token_base64

File: developers/src/assets/js/tools.js (hunk 3 above)
Comment (liuzemei) [reply developers/src/assets/js/tools.js:162]: yes

**Pull request merged**